subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/versionupgrade/MergePullRequest.java

```diff
@@ -0,0 +1,94 @@
+package org.shipkit.internal.gradle.versionupgrade;
+
+import java.io.IOException;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.json.simple.JsonArray;
+import org.json.simple.JsonObject;
+import org.json.simple.Jsoner;
+import org.shipkit.internal.gradle.util.BranchUtils;
+import org.shipkit.internal.util.GitHubApi;
+import org.shipkit.internal.util.IncubatingWarning;
+
+class MergePullRequest {
+
+    private static final Logger LOG = Logging.getLogger(MergePullRequest.class);
+
+    public void mergePullRequest(MergePullRequestTask task) throws IOException {
+        mergePullRequest(task, new GitHubApi(task.getGitHubApiUrl(), task.getAuthToken()));
+    }
+
+    public void mergePullRequest(MergePullRequestTask task, GitHubApi gitHubApi) throws IOException {
+        if (task.isDryRun()) {
+            LOG.lifecycle(" Skipping pull request merging due to dryRun = true");
+            return;
+        }
+
+        String headBranch = BranchUtils.getHeadBranch(task.getForkRepositoryName(), task.getVersionBranch());
+
+        IncubatingWarning.warn("merge pull requests");
+        LOG.lifecycle("Waiting for status of a pull request in repository '{}' between base = '{}' and head = '{}'.", task.getUpstreamRepositoryName(), task.getVersionUpgrade().getBaseBranch(), headBranch);
+
+        String sha = retrievePullRequestSha(task, gitHubApi, headBranch);
+        boolean isPending = true;
+        String body = "{" +
+            "  \"head\": \"" + headBranch + "\"," +
+            "  \"base\": \"" + task.getVersionUpgrade().getBaseBranch() + "\"" +
+            "}";
+
+        int timeouts = 0;
+
+        try {
+            while (isPending) {
+                if (timeouts > 20) {
+                    throw new RuntimeException("To many retries. Merge aborted");
+                }
+
+                String statusesResponse = gitHubApi.get("/repos/" + task.getUpstreamRepositoryName() + "/statuses/" + sha);
+                JsonArray statuses = Jsoner.deserialize(statusesResponse, new JsonArray());
+
+                for (Object status : statuses) {
+                    String state = ((JsonObject) status).getString("state");
+                    String description = ((JsonObject) status).getString("description");
+                    isPending = stateResolver(state, description);
```

Comment:
What if the first check is pending while the second one passed? We should wait for all the checks to pass before merging, so it should be a logical conjunction instead of just an assignment.

Reply:
It seems that if we use a different endpoint for combined status (/status instead of /statuses), we don't need to iterate over a collection. Also it seems that even if all checks pass we get some additional pending ones for some reason. See these ones (one of commits from shipkit):
https://api.github.com/repos/mockito/shipkit/commits/68f9450/statuses
https://api.github.com/repos/mockito/shipkit/commits/68f9450/status

Reply:
@wwilk To first comment - we will. If any of check will fail, exception will be thrown - loop will stop, GradleException will be rethrown. In case of any Status check in pending state, method will return false so while loop will go once again.

Reply:
Yeah but in case you have a statuses like that:
{ status: 'Pending'}
{ status: 'Success'}

You start with isPending = true.
After the first status you set it to true.
After the second one to false, because it's a success.
So in the end isPending is false, although there is still one check pending.
Or am I missing sth?

But anyway if we just check combined status we don't need to worry about that :)

Reply:
Yeah, your point is true and I don't want why I fixed myself on this solution. It's a stupid miss from me :) I'll change it to the combined endpoint